zerver/views/messages.py

```diff
@@ -263,6 +263,24 @@ def by_pm_with(self, query, operand, maybe_negate):
             return query.where(maybe_negate(cond))
 
     def by_search(self, query, operand, maybe_negate):
+        if settings.USING_PGROONGA:
+            return self._by_search_pgroonga(query, operand, maybe_negate)
+        else:
+            return self._by_search_tsearch(query, operand, maybe_negate)
+
+    def _by_search_pgroonga(self, query, operand, maybe_negate):
+        match_positions_byte = func.pgroonga.match_positions_byte
+        query_extract_keywords = func.pgroonga.query_extract_keywords
+        keywords = query_extract_keywords(operand)
+        query = query.column(match_positions_byte(column("rendered_content"),
+                                                  keywords).label("content_matches"))
+        query = query.column(match_positions_byte(column("subject"),
+                                                  keywords).label("subject_matches"))
+        condition = or_(column("subject").op("@@")(operand),
+                        column("rendered_content").op("@@")(operand))
+        return query.where(maybe_negate(condition))
```

Comment:
Sure. I've added a [new function](http://pgroonga.github.io/reference/functions/pgroonga-match-positions-byte.html) that compatible with `ts_match_locs_array` in tsearch_extras to PGroonga.

We can use the existing code for highlighting.